File: python/src/rag_system/knowledge_base_loader.py
"""
Knowledge Base Loader Utility
Loads knowledge documents from JSON files into the Knowledge Base Manager
"""

import json
import os
from pathlib import Path
from typing import Dict, List, Any
from datetime import datetime

from .knowledge_base_manager import KnowledgeBaseManager, KnowledgeBaseCategory
from ..models.knowledge_document import (
    KnowledgeDocument,
    AuthoritySource,
    DocumentMetadata,
    CitationInfo,
    ValidationStatus,
    MedicalDomain
)


class KnowledgeBaseLoader:
    """Loads knowledge documents from JSON files into the Knowledge Base Manager"""

    def __init__(self, manager: KnowledgeBaseManager, data_path: str = "data/knowledge-base"):
        """
        Initialize the loader.
        
        Args:
            manager: Knowledge base manager instance
            data_path: Path to the knowledge base data directory
        """
        self.manager = manager
        self.data_path = Path(data_path)

    def load_all(self) -> Dict[str, Any]:
        """
        Load all documents from the knowledge base directory.
        
        Returns:
            Dictionary with loading results including counts and errors
        """
        results = {
            "loaded": 0,
            "failed": 0,
            "errors": []
        }

        categories = [
            ("clinical-guidelines", KnowledgeBaseCategory.CLINICAL_GUIDELINES),
            ("educational-content", KnowledgeBaseCategory.EDUCATIONAL_CONTENT),
            ("drug-information", KnowledgeBaseCategory.DRUG_INFORMATION),
            ("care-protocols", KnowledgeBaseCategory.CARE_PROTOCOLS),
        ]

        for dir_name, category in categories:
            category_path = self.data_path / dir_name

            if not category_path.exists():
                results["errors"].append(f"Directory not found: {category_path}")
                continue

            json_files = list(category_path.glob("*.json"))

            for file_path in json_files:
                try:
                    result = self.load_document(str(file_path), category)
                    if result["success"]:
                        results["loaded"] += 1
                    else:
                        results["failed"] += 1
                        results["errors"].append(f"{file_path.name}: {result.get('error', 'Unknown error')}")
                except Exception as e:
                    results["failed"] += 1
                    results["errors"].append(f"{file_path.name}: {str(e)}")

        return results

    def load_document(
        self,
        file_path: str,
        category: KnowledgeBaseCategory
    ) -> Dict[str, Any]:
        """
        Load a single document from a JSON file.
        
        Args:
            file_path: Path to the JSON file
            category: Knowledge base category
            
        Returns:
            Dictionary with success status and optional error message
        """
        try:
            with open(file_path, 'r', encoding='utf-8') as f:
                json_doc = json.load(f)

            # Convert JSON document to KnowledgeDocument
            document = KnowledgeDocument(
                document_id=json_doc["documentId"],
                source=AuthoritySource(
                    organization_name=json_doc["source"]["organizationName"],
                    organization_type=json_doc["source"]["organizationType"],
                    url=json_doc["source"].get("url"),
                    credibility_score=json_doc["source"]["credibilityScore"]
                ),
                domain=MedicalDomain(json_doc["domain"]),
                content=json_doc["content"],
                metadata=DocumentMetadata(
                    title=json_doc["metadata"]["title"],
                    publication_date=datetime.fromisoformat(
                        json_doc["metadata"]["publicationDate"].replace('Z', '+00:00')
                    ),
                    version=json_doc["metadata"].get("version"),
                    keywords=json_doc["metadata"]["keywords"],
                    language=json_doc["metadata"]["language"],
                    document_type=json_doc["metadata"]["documentType"]
                ),
                last_updated=datetime.fromisoformat(
                    json_doc["lastUpdated"].replace('Z', '+00:00')
                ),
                validation_status=self._parse_validation_status(json_doc["validationStatus"]),
                citation_format=CitationInfo(
                    citation_text=json_doc["citationFormat"]["citationText"],
                    citation_style=json_doc["citationFormat"]["citationStyle"],
                    url=json_doc["citationFormat"].get("url")
                )
            )

            added = self.manager.add_document(document, category)

            if not added:
                return {
                    "success": False,
                    "error": "Document already exists or could not be added"
                }

            return {"success": True}

        except Exception as e:
            return {
                "success": False,
                "error": str(e)
            }

    def load_category(self, category: KnowledgeBaseCategory) -> Dict[str, Any]:
        """
        Load documents from a specific category.
        
        Args:
            category: Knowledge base category to load
            
        Returns:
            Dictionary with loading results
        """
        results = {
            "loaded": 0,
            "failed": 0,
            "errors": []
        }

        category_dirs = {
            KnowledgeBaseCategory.CLINICAL_GUIDELINES: "clinical-guidelines",
            KnowledgeBaseCategory.EDUCATIONAL_CONTENT: "educational-content",
            KnowledgeBaseCategory.DRUG_INFORMATION: "drug-information",
            KnowledgeBaseCategory.CARE_PROTOCOLS: "care-protocols",
        }

        category_path = self.data_path / category_dirs[category]

        if not category_path.exists():
            results["errors"].append(f"Directory not found: {category_path}")
            return results

        json_files = list(category_path.glob("*.json"))

        for file_path in json_files:
            try:
                result = self.load_document(str(file_path), category)
                if result["success"]:
                    results["loaded"] += 1
                else:
                    results["failed"] += 1
                    results["errors"].append(f"{file_path.name}: {result.get('error', 'Unknown error')}")
            except Exception as e:
                results["failed"] += 1
                results["errors"].append(f"{file_path.name}: {str(e)}")

        return results

    def _parse_validation_status(self, status: str) -> ValidationStatus:
        """Parse validation status string to enum"""
        status_lower = status.lower()
        if status_lower == "validated":
            return ValidationStatus.VALIDATED
        elif status_lower == "pending_review":
            return ValidationStatus.PENDING_REVIEW
        elif status_lower == "outdated":
            return ValidationStatus.OUTDATED
        else:
            return ValidationStatus.PENDING_REVIEW


def create_and_load_knowledge_base(data_path: str = None) -> KnowledgeBaseManager:
    """
    Convenience function to create and load a knowledge base.
    
    Args:
        data_path: Optional path to knowledge base data directory
        
    Returns:
        Loaded knowledge base manager
    """
    manager = KnowledgeBaseManager()
    loader = KnowledgeBaseLoader(
        manager,
        data_path if data_path else "data/knowledge-base"
    )
    results = loader.load_all()

    print(f"Knowledge Base loaded: {results['loaded']} documents loaded, {results['failed']} failed")

    if results["errors"]:
        print("Errors during loading:")
        for error in results["errors"]:
            print(f"  - {error}")

    return manager

File: src/knowledge-context-layer/knowledge-base-loader.ts
/**
 * Knowledge Base Loader Utility
 * Loads knowledge documents from JSON files into the Knowledge Base Manager
 */

import * as fs from 'fs';
import * as path from 'path';
import { KnowledgeBaseManager } from './knowledge-base-manager';
import { KnowledgeDocument, ValidationStatus, MedicalDomain } from '../shared/types/knowledge-document';
import { KnowledgeBaseCategory } from './types';

interface DocumentSource {
  name: string;
  url: string;
  authority: string;
  publicationDate: string;
}

interface DocumentMetadata {
  version: string;
  keywords: string[];
  targetAudience: string;
  [key: string]: any;
}

interface CitationFormat {
  apa: string;
  shortForm: string;
}

interface JsonDocument {
  documentId: string;
  source: DocumentSource;
  domain: string;
  title: string;
  content: string;
  metadata: DocumentMetadata;
  lastUpdated: string;
  validationStatus: string;
  citationFormat: CitationFormat;
}

export class KnowledgeBaseLoader {
  private manager: KnowledgeBaseManager;
  private dataPath: string;

  constructor(manager: KnowledgeBaseManager, dataPath: string = 'data/knowledge-base') {
    this.manager = manager;
    this.dataPath = dataPath;
  }

  /**
   * Load all documents from the knowledge base directory
   */
  loadAll(): { loaded: number; failed: number; errors: string[] } {
    const results = {
      loaded: 0,
      failed: 0,
      errors: [] as string[],
    };

    const categories = [
      { dir: 'clinical-guidelines', category: KnowledgeBaseCategory.CLINICAL_GUIDELINES },
      { dir: 'educational-content', category: KnowledgeBaseCategory.EDUCATIONAL_CONTENT },
      { dir: 'drug-information', category: KnowledgeBaseCategory.DRUG_INFORMATION },
      { dir: 'care-protocols', category: KnowledgeBaseCategory.CARE_PROTOCOLS },
    ];

    for (const { dir, category } of categories) {
      const categoryPath = path.join(this.dataPath, dir);
      
      if (!fs.existsSync(categoryPath)) {
        results.errors.push(`Directory not found: ${categoryPath}`);
        continue;
      }

      const files = fs.readdirSync(categoryPath).filter((f) => f.endsWith('.json'));

      for (const file of files) {
        const filePath = path.join(categoryPath, file);
        try {
          const result = this.loadDocument(filePath, category);
          if (result.success) {
            results.loaded++;
          } else {
            results.failed++;
            results.errors.push(`${file}: ${result.error}`);
          }
        } catch (error) {
          results.failed++;
          results.errors.push(`${file}: ${error instanceof Error ? error.message : 'Unknown error'}`);
        }
      }
    }

    return results;
  }

  /**
   * Load a single document from a JSON file
   */
  loadDocument(
    filePath: string,
    category: KnowledgeBaseCategory
  ): { success: boolean; error?: string } {
    try {
      const fileContent = fs.readFileSync(filePath, 'utf-8');
      const jsonDoc: JsonDocument = JSON.parse(fileContent);

      // Convert JSON document to KnowledgeDocument
      const document: KnowledgeDocument = {
        documentId: jsonDoc.documentId,
        source: {
          name: jsonDoc.source.name,
          url: jsonDoc.source.url,
          authority: jsonDoc.source.authority,
          publicationDate: new Date(jsonDoc.source.publicationDate),
        },
        domain: jsonDoc.domain as MedicalDomain,
        content: jsonDoc.content,
        metadata: {
          title: jsonDoc.title,
          version: jsonDoc.metadata.version,
          keywords: jsonDoc.metadata.keywords,
          ...jsonDoc.metadata,
        },
        lastUpdated: new Date(jsonDoc.lastUpdated),
        validationStatus: this.parseValidationStatus(jsonDoc.validationStatus),
        citationFormat: jsonDoc.citationFormat,
      };

      const added = this.manager.addDocument(document, category);

      if (!added) {
        return { success: false, error: 'Document already exists or could not be added' };
      }

      return { success: true };
    } catch (error) {
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Unknown error',
      };
    }
  }

  /**
   * Load documents from a specific category
   */
  loadCategory(category: KnowledgeBaseCategory): { loaded: number; failed: number; errors: string[] } {
    const results = {
      loaded: 0,
      failed: 0,
      errors: [] as string[],
    };

    const categoryDirs: Record<KnowledgeBaseCategory, string> = {
      [KnowledgeBaseCategory.CLINICAL_GUIDELINES]: 'clinical-guidelines',
      [KnowledgeBaseCategory.EDUCATIONAL_CONTENT]: 'educational-content',
      [KnowledgeBaseCategory.DRUG_INFORMATION]: 'drug-information',
      [KnowledgeBaseCategory.CARE_PROTOCOLS]: 'care-protocols',
    };

    const categoryPath = path.join(this.dataPath, categoryDirs[category]);

    if (!fs.existsSync(categoryPath)) {
      results.errors.push(`Directory not found: ${categoryPath}`);
      return results;
    }

    const files = fs.readdirSync(categoryPath).filter((f) => f.endsWith('.json'));

    for (const file of files) {
      const filePath = path.join(categoryPath, file);
      try {
        const result = this.loadDocument(filePath, category);
        if (result.success) {
          results.loaded++;
        } else {
          results.failed++;
          results.errors.push(`${file}: ${result.error}`);
        }
      } catch (error) {
        results.failed++;
        results.errors.push(`${file}: ${error instanceof Error ? error.message : 'Unknown error'}`);
      }
    }

    return results;
  }

  private parseValidationStatus(status: string): ValidationStatus {
    switch (status.toLowerCase()) {
      case 'validated':
        return ValidationStatus.VALIDATED;
      case 'pending_review':
        return ValidationStatus.PENDING_REVIEW;
      case 'outdated':
        return ValidationStatus.OUTDATED;
      default:
        return ValidationStatus.PENDING_REVIEW;
    }
  }
}

/**
 * Convenience function to create and load a knowledge base
 */
export function createAndLoadKnowledgeBase(dataPath?: string): KnowledgeBaseManager {
  const manager = new KnowledgeBaseManager();
  const loader = new KnowledgeBaseLoader(manager, dataPath);
  const results = loader.loadAll();

  console.log(`Knowledge Base loaded: ${results.loaded} documents loaded, ${results.failed} failed`);
  
  if (results.errors.length > 0) {
    console.error('Errors during loading:');
    results.errors.forEach((error) => console.error(`  - ${error}`));
  }

  return manager;
}
